### PR TITLE
feat(pxp): add zephyr support

### PR DESCRIPTION
**Pull request by faxe1008**
Add semaphore and connect IRQ.
Originally proposed by @0xFarahFl. [https://github.com/zephyrproject-rtos/lvgl/pull/52](https://github.com/zephyrproject-rtos/lvgl/pull/52)

### Description of the feature or fix

A clear and concise description of what the bug or new feature is.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.12](https://github.com/szepeviktor/astyle/releases/tag/v3.4.12) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Comment by kisvegabor** [comment]
Hi,

We can merge it now as a hotfix, however I think a proper solution would be to
1. Add built-in Zephyr support to LVGL. See [here](https://github.com/lvgl/lvgl/tree/master/src/osal)
2. Use LVGL's OS API in PXP too.

What do you think?

cc @nicusorcitu @anaGrad

**Comment by faxe1008** [comment]
Hey @kisvegabor 

> Add built-in Zephyr support to LVGL. See [here](https://github.com/lvgl/lvgl/tree/master/src/osal)
With that one I am still a bit hesitant - the OSAL needs dynamic thread creation and this is still marked as experimental so subject to change. By keeping the OSAL intree for now we can make sure that any changes will cause CI breakage and are addresed. [https://github.com/zephyrproject-rtos/zephyr/blob/74ddfbc8a2e4f1932bf2b6a160f83a27a6c135a5/kernel/Kconfig#L237](https://github.com/zephyrproject-rtos/zephyr/blob/74ddfbc8a2e4f1932bf2b6a160f83a27a6c135a5/kernel/Kconfig#L237)

> Use LVGL's OS API in PXP too.

Good point I'll adjust this PR accordingly :^)

**Review by kisvegabor — APPROVED** [review]
Thank you! It's good for me, but I hope @nicusorcitu and/or @anaGrad can take a look too.

**Comment by nicusorcitu** [comment]
@cristian-stoica should review it also

File: src/draw/nxp/pxp/lv_pxp_osa.c [hunk 1]
```diff
@@ -66,11 +66,8 @@ static void _pxp_wait(void);
  *  STATIC VARIABLES
  **********************/
 
-#if defined(SDK_OS_FREE_RTOS)
-    static SemaphoreHandle_t xPXPIdleSemaphore;
-#endif
-#if defined(__ZEPHYR__)
-    static K_SEM_DEFINE(s_pxpIdleSem, 0, 1);
+#if LV_USE_OS
```
Comment (nicusorcitu) [inline src/draw/nxp/pxp/lv_pxp_osa.c:69]: PXP can work with or without FreeRTOS support.
PXP in baremetal mode means that SDK_OS_FREE_RTOS = 0.

While LV_USE_OS is a feature added in LVGL v9 for drawing parallelism. Those two defines have a different meaning.
I will give you an example:
```
SDK_OS_FREE_RTOS = 1
LV_USE_OS = LV_OS_NONE
``` 
This means that we will use LVGL single thread mode (same as LVGL v8) but PXP will have FreeRTOS support.

**Review by FASTSHIFT — APPROVED** [review]

File: src/draw/nxp/pxp/lv_pxp_osa.c (hunk 1 above)
Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c:69]: Understood, so this destinction needs to be handled.

**Comment by 0xFarahFl** [comment]
Sorry, it might not be directly related to this PR but my question concerns the lvgl_support.h in this [PR ](https://github.com/zephyrproject-rtos/lvgl/blame/zephyr/src/draw/nxp/pxp/lv_draw_pxp_blend.c). In order to test the zephyr support, I had to comment it and add LCD_WIDTH and LCD_HEIGHT macros but I want to know what was the intended content of this file and where to find it?

**Comment by nicusorcitu** [comment]
LCD_WIDTH and LCD_HEIGHT were used in LVGL v8 just to define a temporary buffer needed by PXP for processing complex operations into multiple steps.
The file "lvgl_support.h" can be find in the SDK releases on official website at NXP.

That buffer has been removed in LVGl v9 so the defines are no longer needed in LVGL v9.

**Comment by 0xFarahFl** [comment]
Can you provide me with the link please?

**Comment by cristian-stoica** [comment]
> Can you provide me with the link please?

https://github.com/nxp-mcuxpresso/lvgl is the repository for NXP LVGL releases. 
They are integrated in the SDKs which you can configure and download at https://www.nxp.com/design/design-center/software/development-software/mcuxpresso-software-and-tools-/mcuxpresso-software-development-kit-sdk:MCUXpresso-SDK

**Comment by nicusorcitu** [comment]
Hi @FASTSHIFT
The PR was not yet complete, some behaviors have changed, we need to get back on this one.
Thanks,
Nicu.

**Comment by FASTSHIFT** [comment]
> Hi @FASTSHIFT The PR was not yet complete, some behaviors have changed, we need to get back on this one. Thanks, Nicu.

Ah my bad, I revert it https://github.com/lvgl/lvgl/pull/6261

File: src/draw/nxp/pxp/lv_pxp_osa.c [hunk 2]
```diff
@@ -25,6 +25,11 @@
     #include "semphr.h"
 #endif
 
+#if defined(__ZEPHYR__)
```
Comment (nicusorcitu) [inline src/draw/nxp/pxp/lv_pxp_osa.c:28]: Please remove the above inclusions:
```
#if defined(SDK_OS_FREE_RTOS)
    #include "FreeRTOS.h"
    #include "semphr.h"
#endif
```

and replace it with:
```
#include "../../../osal/lv_os.h"
```

File: src/draw/nxp/pxp/lv_pxp_osa.c (hunk 1 above)
Comment (nicusorcitu) [reply src/draw/nxp/pxp/lv_pxp_osa.c:69]: We can fix later this behavior. If you fix-up the inclusions then you can merge this PR as is.

File: src/draw/nxp/pxp/lv_pxp_osa.c (hunk 2 above)
Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c:28]: Applied your suggestion.

Comment (nicusorcitu) [reply src/draw/nxp/pxp/lv_pxp_osa.c:28]: where?

Comment (nicusorcitu) [reply src/draw/nxp/pxp/lv_pxp_osa.c:28]: Just need the include and then the PR can be merged.

Comment (faxe1008) [reply src/draw/nxp/pxp/lv_pxp_osa.c:28]: I suppose that since the PR was merged once GitHub does not display new changes. I guess a new PR is needed

**Pull request merged**